Adds docs for notRestoredReasons (#27947)
mdn/content: files/en-us/web/api/performance_api/reporting_backforward_cache_not_restored_reasons/index.md

---
title: Reporting back/forward cache not restored reasons
slug: Web/API/Performance_API/Reporting_backforward_cache_not_restored_reasons
page-type: web-api-overview
status:
  - experimental
browser-compat: api.PerformanceNavigationTiming.notRestoredReasons
---

{{DefaultAPISidebar("Performance API")}}{{SeeCompatTable}}

The {{domxref("PerformanceNavigationTiming.notRestoredReasons")}} property reports information on whether frames present in the document were blocked from using the [back/forward cache](https://web.dev/bfcache/) (bfcache) on navigation, and why. Developers can use this information to identify pages that need updates to make them bfcache-compatible, thereby improving site performance.

## Concepts and usage

Modern browsers provide an optimization feature for history navigation called the [back/forward cache](https://web.dev/bfcache/) (bfcache). This enables an instant loading experience when users go back to a page they have already visited. Pages can be blocked from entering the bfcache or get evicted while in the bfcache for different reasons, some required by a specification and some specific to browser implementations.

> Previously, there was no way for developers to find out why their pages were blocked from using the bfcache in the wild, though there is a [test in Chrome dev tools](https://web.dev/bfcache/#test-to-ensure-your-pages-are-cacheable). To enable monitoring in the field, the [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) class has been extended to include a `notRestoredReasons` property. This returns an object containing related information on all frames present in the document:

Reviewer: Currently reads a bit like a browser explaining why they are launching this API :-)

```suggestion
Previously, there was no way for developers to find out why their pages were blocked from using the bfcache in the wild, though there is a [test in Chrome dev tools](https://web.dev/bfcache/#test-to-ensure-your-pages-are-cacheable). To enable monitoring in the field use the `notRestoredReasons` property of the [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) class. This returns an object containing related information on all frames present in the document:
```

- Whether they were blocked from using the bfcache.
- Reasons why they were blocked from using the bfcache.

This allows developers to take action to make those pages bfcache-compatible, thereby improving site performance.

## Examples

A [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) instance can be obtained from features such as [`Performance.getEntriesByType()`](/en-US/docs/Web/API/Performance/getEntriesByType) and [`PerformanceObserver`](/en-US/docs/Web/API/PerformanceObserver).

For example, you could invoke the following function to return all `PerformanceNavigationTiming` objects currently present in the performance timeline and log their `notRestoredReasons`:

```js
function returnNRR() {
  const navEntries = performance.getEntriesByType("navigation");
  for (let i = 0; i < navEntries.length; i++) {
    console.log(`Navigation entry ${i}`);
    let navEntry = navEntries[i];
    console.log(navEntry.notRestoredReasons);
  }
}
```

The `PerformanceNavigationTiming.notRestoredReasons` property returns an object with the following structure, which represents the blocked state of the top-level frame:

> ```js

Reviewer: ```suggestion
```json
```

Reviewer 2: This is not JSON. I'd say get rid of this code altogether since the prose below describes the exact same thing in better detail.

Author: Yeah, my understanding is that it is a JS object, but not actually JSON. I think we still need an example mind; it would read a bit strange to define what the object properties are without showing the actual object.

Reviewer: Ah yes, sorry forgot this is returned as JavaScript API rather than reported (e.g. like a CSP exception) so it is indeed JavaScript not JSON. I agree an example would be useful IMHO.

- `reasons`
  - : An array of strings each representing a reason why the navigated page was blocked from using the bfcache. There are many different reasons why blocking could occur. See the [Blocking reasons](#blocking_reasons) for more details.
- `src`
  - : A string representing the path to the frame's source (for example `<iframe src="b.html">`). If the frame has no `src`, the value will be an empty string.
- `url`
  - : A string representing the URL of the navigated page.

### Reporting bfcache blocking in same-origin frames

When a page has same-origin frames embedded, the returned `notRestoredReasons` value will contain an object inside the `children` property representing the blocked state of each embedded frame.

For example:

```js
{
  blocked: false,
  children: [
    { url: "a.com", src: "b.a.com", id: "b", name: "b", blocked: false, reasons: [], children: [] },
    { url: "a.com", src: "c.a.com", id: "c", name: "c", blocked: true, reasons: [ "BroadcastChannel" ], children: [] },
    { url: "a.com", src: "d.a.com", id: "d", name: "d", blocked: false, reasons: [], children: [] }
  ],
  id: "",
  name: "",
  reasons: [],
  src: "",
  url:"a.com"

> }

Reviewer 2: For cases like this, I suggest either making it actual valid JSON, or make it valid JS, like by adding parentheses: `({...});`

Author: I think this is reasonable. Updated.

```js
{
  blocked: false,
  children: [
    { url: "a.com", src: "c.a.com", id: "c", name: "c", blocked: true, reasons: [ "ScreenReader" ], children: [] },
    /* cross-origin frame */
    { url: "", src: "b.com", id: "d", name: "d", blocked: true, reasons: [], children: [] }
  ],
  id: "",
  name: "",
  reasons: [],
  src: "",
  url:"a.com"
}
```

If multiple cross-origin frames have blocking reasons, we randomly select one cross-origin frame and report whether it blocked bfcache or not. For the rest of the frames, we report `null` for the `blocked` value. This is to stop bad actors from inferring information about user state on sites they don't control by embedding multiple third-party frames into a page and then comparing the blocking information from each.

```js
{
  blocked: false,
  children: [
    /* cross-origin frames */
    {url: "", src: "b.com", id: "b", name: "b", blocked: null, reasons: [], children: []},
    {url: "", src: "c.com", id: "c", name: "c", blocked: true, reasons: [], children: []},
    {url: "", src: "d.com", id: "d", name: "d", blocked: null, reasons: [], children: []}
  ]
  id: "",
  name: "",
  reasons: [],
  src: "",
  url:"a.com"
}
```

> > **Note:** See the [Security and privacy](https://github.com/rubberyuzu/bfcache-not-retored-reason/blob/main/NotRestoredReason.md#security-and-privacy) section in the explainer for more details about security and privacy considerations.

Reviewer: Should the alient points be added in MDN here rather than referring to an explainer for launching the product?

Author: Yeah, I think you've got a good point here. I've removed the explainer link and added a few bits that were not already mentioned in the page. TBH, most of it was covered already.

## Blocking reasons

> As noted earlier, there are many different reasons why blocking could occur. Google has compiled a [spreadsheet](https://docs.google.com/spreadsheets/d/1li0po_ETJAIybpaSX5rW_lUN62upQhY0tH4pR5UPt60/edit#gid=0) showing all the reason strings and explaining what they mean.

Reviewer: I think we should remove this line (or at the very least clarify these are Chrome's reasons).

AFAIK the reasons are not standard and so would be browser specific. We should probably note that?

Author: Yup, good point. Updated.

> There are a few major categories of reasons that are worth calling out:

Reviewer: ```suggestion
As an example, Chrome's implementation has a few major categories of reasons:
```

- `SupportPending`: The developer's code is using a web platform feature that disqualifies the page from the bfcache, for example the [Web Serial API](/en-US/docs/Web/API/Web_Serial_API), [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API), [File System Access API](/en-US/docs/Web/API/File_System_Access_API), or [Media Session API](/en-US/docs/Web/API/Media_Session_API). Or the page is using [`Cache-Control: no-store`](/en-US/docs/Web/HTTP/Headers/Cache-Control), which currently [prevents the bfcache from being used in some browsers](https://web.dev/bfcache/#minimize-use-of-cache-control-no-store). This category is also used to report the presence of a tool outside the page itself that is blocking the bfcache, such as a screen reader or password manager.

## Browser compatibility

> {{Compat}}

Reviewer: Should there be a Specifications section pointing to: https://github.com/WICG/bfcache-not-restored-reason/blob/main/NotRestoredReason.md ?

Reviewer: Well I agree it's not a "standard", but it's the nearest thing we have to a "specification" for now. Not sure what the precedent for this sort of pre-standard thing is on MDN? Or maybe it usually has a spec by the time we add to here, and this one is just a little behind?

At the very least, I think this link should be listed in the "See also" section.

Author: I'll add it to the "See also" section. Prestandard stuff doesn't get added to MDN very often, but I thought this one would be OK as it is just a single property, and it is close to standard. At worst, I might have to mark it as non-standard until it actually gets added to the spec.

---
title: Reporting back/forward cache not restored reasons
slug: Web/API/Performance_API/Reporting_backforward_cache_not_restored_reasons
page-type: web-api-overview
status:
  - experimental
browser-compat: api.PerformanceNavigationTiming.notRestoredReasons
---

{{DefaultAPISidebar("Performance API")}}{{SeeCompatTable}}

> The {{domxref("PerformanceNavigationTiming.notRestoredReasons")}} property reports information on whether frames present in the document were blocked from using the [back/forward cache](https://web.dev/bfcache/) (bfcache) on navigation, and why. Developers can use this information to identify pages that need updates to make them bfcache-compatible, thereby improving site performance.

Author: updated

## Concepts and usage

Modern browsers provide an optimization feature for history navigation called the [back/forward cache](https://web.dev/bfcache/) (bfcache). This enables an instant loading experience when users go back to a page they have already visited. Pages can be blocked from entering the bfcache or get evicted while in the bfcache for different reasons, some required by a specification and some specific to browser implementations.

To enable monitoring bfcache blocking in the field the [`PerformanceNavigationTiming`](/en-US/docs/Web/API/PerformanceNavigationTiming) class includes a `notRestoredReasons` property. This returns an object containing related information on the top-level frame and all {{htmlelement("iframe")}}s present in the document:

> - Reasons why they were blocked from using the bfcache.

Author: updated

For example, you could invoke the following function to return all `PerformanceNavigationTiming` objects currently present in the performance timeline and log their `notRestoredReasons`:

```js
function returnNRR() {
  const navEntries = performance.getEntriesByType("navigation");
  for (let i = 0; i < navEntries.length; i++) {
    console.log(`Navigation entry ${i}`);
    let navEntry = navEntries[i];
    console.log(navEntry.notRestoredReasons);
  }
}
```

For history navigations, the {{domxref("PerformanceNavigationTiming.notRestoredReasons")}} property returns a {{domxref("NotRestoredReasons")}} object with the following structure, which represents the blocked state of the top-level frame:

```js
{
  children: [],
  id: null,
  name: null,
  reasons: [
    { reason: "unload-listener" }
  ],
  src: "",
  url: "example.com",
}
```

The properties are as follows:

- {{domxref("NotRestoredReasons.children", "children")}} {{ReadOnlyInline}} {{Experimental_Inline}}

>   - : An array of {{domxref("NotRestoredReasons")}} objects representing the blocked state of any child {{htmlelement("iframe")}}s embedded in the current document. Each object has the same structure as the parent object — this way, any number of levels of embedded `<iframe>s` can be represented inside the object recursively. If the frame has no children, the array will be empty; if the document is in a cross-origin `<iframe>`, `children` will return `null`.

Author: I updated this in the same way as on the `NotRestoredReasons` page. See above.

- {{domxref("NotRestoredReasons.id", "id")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the `id` attribute value of the `<iframe>` the document is contained in (for example `<iframe id="foo" src="...">`). If the document is not in an `<iframe>` or the `<iframe>` has no `id` set, `id` will return `null`.
- {{domxref("NotRestoredReasons.name", "name")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the `name` attribute value of the `<iframe>` the document is contained in (for example `<iframe name="bar" src="...">`). If the document is not in an `<iframe>` or the `<iframe>` has no `name` set, `name` will return `null`.
- {{domxref("NotRestoredReasons.reasons", "reasons")}} {{ReadOnlyInline}} {{Experimental_Inline}}

>   - : An array of {{domxref("NotRestoredReasonDetails")}} objects, each representing a reason why the navigated page was blocked from using the bfcache. If the document is in a cross-origin `<iframe>`, `reasons` will return `null`. See [Blocking reasons](#blocking_reasons) for more details on the reasons.

Author: Updated

- {{domxref("NotRestoredReasons.src", "src")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the path to the source of the `<iframe>` the document is contained in (for example `<iframe src="exampleframe.html">`). If the document is not in an `<iframe>`, `src` will return `null`.
- {{domxref("NotRestoredReasons.url", "url")}} {{ReadOnlyInline}} {{Experimental_Inline}}
  - : A string representing the URL of the navigated page or `<iframe>`. If the document is in a cross-origin `<iframe>`, `url` will return `null`.

### Reporting bfcache blocking in same-origin `<iframe>`s

> When a page has same-origin `<iframe>`s embedded, the returned `notRestoredReasons` value will contain an array of objects inside the `children` property representing the blocked state of each embedded frame.

Author: I thought about this a bit, and ended up using the following:

> When a page has same-origin `<iframe>`s embedded, the returned `notRestoredReasons` value will contain an array of objects inside the `children` property representing the blocking reasons related to each embedded frame.

For example:

```js
{
  children: [
    {
      children: [],
      id: "iframe-id",
      name: "iframe-name",
      reasons: [],
      src: "./index.html",
      url: "https://www.example.com/iframe-examples.html"
    },
    {
      children: [],
      id: "iframe-id2",
      name: "iframe-name2",
      reasons: [
        { "reason": "unload-listener" }
      ],
      src: "./unload-examples.html",
      url: "https://www.example.com/unload-examples.html"
    },
  ],
  id: null,
  name: null,
  reasons: [],
  src: null,
  url:"https://www.example.com"
}
```

### Reporting bfcache blocking in cross-origin `<iframe>`s

When a page has cross-origin frames embedded, we limit the amount of information shared about them to avoid leaking cross-origin information. We only include information that the outer page already knows, and whether the cross-origin subtree blocked the bfcache or not. We don't include any blocking reasons or information about lower levels of the subtree (even if some sub-levels are same-origin).

For example:

```js
{
  children: [
    {
      children: [],
      id: "iframe-id",
      name: "iframe-name",
      reasons: [],

>       src: "./index.html",

Reviewer: This should be fully qualified if it's a cross-origin iframe (a relative path is, by definition same-origin). Also probably can drop the `index.html`.

```suggestion
      src: "https://www.example2.com/",
```

Author: Ooops, yes, updated.

>       url: "https://www.example2.com/"

Reviewer: I think this should be null for cross-origin? It starts the same as the src, but could diverge but we won't know.

```suggestion
      url: "https://www.example2.com/"
```
```suggestion
      url: null
```

Author: Yup, that's right. This also agrees with the notes on the `url` reference page. Updated.

- `"lock"`
  - : While unloading, held locks and lock requests were terminated, so the page was not in a stable state that could be stored in the bfcache.
- `"masked"`
  - : The exact reason is hidden for privacy purposes. This value can mean one of the following:
    - The current document has children contained in a cross-origin {{htmlelement("iframe")}}, and they prevented storage in the bfcache.
    - The current Document could not be stored in the bfcache for user agent-specific reasons.
- `"navigation-failure"`
  - : The original navigation that created the current document errored, and storing the resulting error document in the bfcache was prevented.
- `"parser-aborted"`
  - : The current document never finished its initial HTML parsing, and storing the unfinished document in the bfcache was prevented.
- `"websocket"`
  - : While unloading, an open [WebSocket](/en-US/docs/Web/API/WebSockets_API) connect was shut down, so the page was not in a stable state that could be stored in the bfcache.

> Chrome implements the following additional blocking reasons (among others):

Reviewer: Chrome is the only implementer of this so do we need to say this? Also these [are being standardised](https://github.com/whatwg/html/pull/10154)

```suggestion
The following additional blocking reasons (among others) may be used by some browsers:
```

Author: I tried to update this to be as future proof as possible. I updated the line about the first list to:

> The initial values listed in the specification are:

And then I updated this line to:

> Additional blocking reasons may be used by some browsers, for example: 

this will work for now, until some point in the future when we decide we want to list all blocking reasons on MDN.

Reviewer: I hear the “may block” blocking reasons might change a good bit over time anyway so probably best not to list those or might be forever chasing your tail here depending how often they update.
